pytorch_lightning/trainer/trainer.py

```diff
@@ -676,7 +676,12 @@ def _call_and_handle_interrupt(self, trainer_fn: Callable, *args: Any, **kwargs:
             **kwargs: keyword arguments to be passed to `trainer_fn`
         """
         try:
-            return trainer_fn(*args, **kwargs)
+            if isinstance(self.training_type_plugin, DDPSpawnPlugin):
+                spawn_output: _SpawnOutput = self.training_type_plugin.spawn(trainer_fn, *args, **kwargs)
+                self.training_type_plugin._recover_results_in_main_process(spawn_output, self)
```

Comment:
I really like the simplification this offers. I am wary of tying this to DDP Spawn however. This forces everyone to subclass DDPSpawnPlugin and override different parts.

What do you think about offering an AbstractSpawn interace which requires people to implement `spawn` and `recover_results_in_main_process` and which defines the `spawn_output` struct?

Then DDP Spawn could both inherit from TrainingTypePlugin and implement `AbstractSpawn`

Reply:
@ananthsub I was actually thinking in a similar direction, I imagined it with composition rather than inheritance. My idea was to have an (abstract) Executor interface, of which the TTP owns an instance like the checkpoint IO plugin. 
I'm definitely open to discuss such directions in more detail

Reply:
@awaelchli You mean something like an internal launcher? That either calls the script or uses spawn? I think we should do that to make it way more modular.

Reply:
Yes I was thinking something like that.

Reply:
Will you open an issue for tracking this? :)

Reply:
Yes, this is quite exciting. This would enable us to possibly merge ddp, ddp_spawn into ddp.py and provide a popen or spawn executor.

Reply:
Ideally, we would have `self.training_type_plugin.run(trainer_fn, *args, **kwargs)` and this would execute the associated trainer method with the spawn / none executor.

Reply:
Awesome, we will continue discussion and drafting over here: https://github.com/PyTorchLightning/pytorch-lightning/issues/10985